tenders/common/elf.c

```diff
@@ -73,120 +74,155 @@ static ssize_t pread_in_full(int fd, void *buf, size_t count, off_t offset)
-    Elf64_Half ph_i;
     Elf64_Phdr *phdr = NULL;
-    Elf64_Ehdr hdr;
-
-    /* elf entry point (on physical memory) */
-    *p_entry = 0;
-    /* highest byte of the program (on physical memory) */
-    *p_end = 0;
+    Elf64_Ehdr *ehdr = NULL;
+    Elf64_Addr e_entry;                 /* Program entry point */
+    Elf64_Addr e_end;                   /* Highest memory address occupied */
 
     fd_kernel = open(file, O_RDONLY);
     if (fd_kernel == -1)
         goto out_error;
 
-    nbytes = pread_in_full(fd_kernel, &hdr, sizeof(Elf64_Ehdr), 0);
+    ehdr = malloc(sizeof(Elf64_Ehdr));
+    if (ehdr == NULL)
+        goto out_error;
+    nbytes = pread_in_full(fd_kernel, ehdr, sizeof(Elf64_Ehdr), 0);
     if (nbytes < 0)
         goto out_error;
     if (nbytes != sizeof(Elf64_Ehdr))
         goto out_invalid;
-    if (!ehdr_is_valid(&hdr))
+    if (!ehdr_is_valid(ehdr))
         goto out_invalid;
+    /*
+     * e_entry must be non-zero and within range of our memory allocation.
+     */
+    if (ehdr->e_entry == 0 || ehdr->e_entry >= mem_size)
+        goto out_invalid;
+    e_entry = ehdr->e_entry;
 
-    ph_off = hdr.e_phoff;
-    ph_entsz = hdr.e_phentsize;
-    ph_cnt = hdr.e_phnum;
-    ph_size = ph_entsz * ph_cnt;
+    size_t ph_size = ehdr->e_phnum * ehdr->e_phentsize;
     phdr = malloc(ph_size);
     if (!phdr)
         goto out_error;
-    nbytes = pread_in_full(fd_kernel, phdr, ph_size, ph_off);
+    nbytes = pread_in_full(fd_kernel, phdr, ph_size, ehdr->e_phoff);
     if (nbytes < 0)
         goto out_error;
     if (nbytes != ph_size)
         goto out_invalid;
 
     /*
-     * Load all segments with the LOAD directive from the elf file at offset
-     * p_offset, and copy that into p_addr in memory. The amount of bytes
-     * copied is p_filesz.  However, each segment should be given
-     * p_memsz aligned up to p_align bytes on memory.
+     * Load all program segments with the PT_LOAD directive.
      */
-    for (ph_i = 0; ph_i < ph_cnt; ph_i++) {
-        uint8_t *daddr;
-        uint64_t _end;
-        size_t offset = phdr[ph_i].p_offset;
-        size_t filesz = phdr[ph_i].p_filesz;
-        size_t memsz = phdr[ph_i].p_memsz;
-        uint64_t paddr = phdr[ph_i].p_paddr;
-        uint64_t align = phdr[ph_i].p_align;
-        uint64_t result;
-        int prot;
+    e_end = 0;
+    for (Elf64_Half ph_i = 0; ph_i < ehdr->e_phnum; ph_i++) {
+        Elf64_Addr p_vaddr = phdr[ph_i].p_vaddr;
+        Elf64_Xword p_filesz = phdr[ph_i].p_filesz;
+        Elf64_Xword p_memsz = phdr[ph_i].p_memsz;
+        Elf64_Xword p_align = phdr[ph_i].p_align;
+        Elf64_Addr p_vaddr_end;
 
         if (phdr[ph_i].p_type != PT_LOAD)
             continue;
 
-        if ((paddr >= mem_size) || add_overflow(paddr, filesz, result)
-                || (result >= mem_size))
+        if ((p_vaddr >= mem_size)
+                || add_overflow(p_vaddr, p_filesz, p_vaddr_end)
+                || (p_vaddr_end >= mem_size))
```

Comment:
Hmm, re-reading this, shouldn't it be `> mem_size` instead of `>= mem_size`?
Consider the case where someone has a mapping of 2MB - or a single byte, for simplicity:
```c
p_vaddr = 0;
p_filesz = p_memsz = 1;
mem_size = 1;
/* it follows that */
!( p_vaddr >= mem_size ); /* good, within range */
p_vaddr_end = 0 + 1; /* good, no overflow */
p_vaddr_end >= 1; /* rejected since p_vaddr_end == 1 */
```
Now p_vaddr_end is

Reply:
Done in abeca34.